### ISSUE TITLE
Use `GOMEMLIMIT` to guard against OOM kills in controllers - Upgrade go to `1.19`

### ISSUE THREAD

**Issue by alpeb** [linkerd/linkerd2#9947]
# Problem and repro

In some situations the `destination` container sees its memory consumption progressively augment until it gets OOM killed. We have been able to reproduce that by rolling out emojivoto's pods every 30s (here the rollouts started at 11:55):
![Screenshot from 2022-12-05 12-22-22](https://user-images.githubusercontent.com/554287/205702951-3c828885-def9-4e08-85b2-a965b60d7488.png)
(tracking here `container_memory_working_set_bytes{container="destination"}`, which is the metric the OOM killer watches)

We took heap dumps before starting the rollouts and 25m after, and  generated this comparative result:

```bash
# (using edge-22.12.1)
$ k -n linkerd port-forward linkerd-destination-5dd6c5985f-sl2jz 9996 &
$ curl http://localhost:9996/debug/pprof/heap?gc=1 > heap-before.out
$ while true; do kubectl -n emojivoto rollout restart deploy ; sleep 30; done
# wait 10m
$ curl http://localhost:9996/debug/pprof/heap?gc=1 > heap-after.out
$ go tool pprof -png --base heap-before.out heap-after.out
```
![profile001](https://user-images.githubusercontent.com/554287/205704007-9cf3751c-f9d3-468e-8a7b-998fd43be6e6.png)

This shows 68.71% of the allocated memory during this time span was requested by `k8s.io/apimachinery/pkg/watch.(*StreamWatcher).receive`, so not linkerd's fault.

## GC tuning

Lowering `GOGC` down to 1 forces GC to run as frequently as possible, but the curve above just smoothens up while overall memory consumption and trend remains the same.

However, using `GOMEMLIMIT` (introduced in go 1.19) helps out.  If we set it to 35MiB we get this, triggering the rollouts at 14:32:
![Screenshot from 2022-12-05 14-55-56](https://user-images.githubusercontent.com/554287/205731469-3a078627-888d-41f2-8663-3bbc6e097a8b.png)
Before, we went from 26MB up to 42MB during the rollouts, and now from 25MB to 26MB in about the same time. Memory usage still climbs, but not as bad as before.

`GOMEMLIMIT` requires go 1.19, and it's actually recommended to use in containers with a memory limit; according to [A Guide to the Go Garbage Collector](https://tip.golang.org/doc/gc-guide#Memory_limit):
> Do take advantage of the memory limit when the execution environment of your Go program is entirely within your control, and the Go program is the only program with access to some set of resources (i.e. some kind of memory reservation, like a container memory limit).
A good example is the deployment of a web service into containers with a fixed amount of available memory.
In this case, a good rule of thumb is to leave an additional 5-10% of headroom to account for memory sources the Go runtime is unaware of.

## Course of action

Besides bumping go to 1.19 in all the base images, we can add the env var `GOMEMLIMIT` to the manifests only if the helm value `memory.limit` exists (values-ha.yaml sets it for the control plane containers), set to 90% its value.

Ref #8270

**Comment by olix0r** [comment]
> This shows 68.71% of the allocated memory during this time span was requested by `k8s.io/apimachinery/pkg/watch.(*StreamWatcher).receive`, so not linkerd's fault.

I don't draw that conclusion from this data.

Yes, 69% of _allocations_ occur when parsing JSON data from Kubernetes. This isn't really shocking for an application that mostly watches lots of resources. Pprof _doesn't_ tell us how long these allocations last, though. They could be extremely short-lived allocations, in which case, who cares? Or they parsed objects could be held forever (i.e. by Linkerd), which would actually be a leak. Or, it could just as well be that something in the other 31% of allocations is being retained in a way that appears leaky.

pprof is useful to help us fill in our picture of the system, but we should be wary about drawing conclusions based on it alone.

If `GOMEMLIMIT` is something we should be setting in any case, it seems like a good change (so that the runtime knows its actual available memory). But I'm still pretty skeptical that this explains/resolves the underlying issue: while it's good that we can increase GC as memory pressure increases, and this helps us avoid OOMing, we probably don't want to be in that state in general (in fact it seems like something worth alerting on -- a destination controller in that state should probably get more memory so that it doesn't suffer continuous GC).

What we're really trying to understand is: why do we have memory pressure in the first place? Is it really JSON decoding? If so, it seems like the Kubernetes API server (as well as numerous other controllers) would suffer from this as well. What have they done to address this?

How much memory do we need to allocate to avoid this GC pressure?

**Comment by Teko012** [comment]
We still see regular OOM kills in `linkerd-proxy`, was this implemented for that as well?

**Comment by Teko012** [comment]
@kleimkuhler We are on the latest stable `2.12.4`.

**Comment by kleimkuhler** [comment]
Okay it'd be helpful if you can open up a separate issue with a description about what you're seeing and any relevant information. I'd like to keep this issue specific to the ongoing conversation and if `GOMEMLIMIT` is worth adding, but we can meanwhile dig a little more into your issue.

**Comment by risingspiral** [comment]
Moving away from applying GOMEMLIMIT
